extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ServerVertxAsyncFileMessageBodyWriter.java

```diff
@@ -0,0 +1,63 @@
+package io.quarkus.resteasy.reactive.server.runtime;
+
+import java.lang.reflect.Type;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.ext.Provider;
+
+import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
+import org.jboss.resteasy.reactive.server.spi.ResteasyReactiveResourceInfo;
+import org.jboss.resteasy.reactive.server.spi.ServerHttpResponse;
+import org.jboss.resteasy.reactive.server.spi.ServerMessageBodyWriter;
+import org.jboss.resteasy.reactive.server.spi.ServerRequestContext;
+
+import io.quarkus.resteasy.reactive.common.runtime.VertxAsyncFileMessageBodyWriter;
+import io.vertx.core.Handler;
+import io.vertx.core.file.AsyncFile;
+
+@Provider
+public class ServerVertxAsyncFileMessageBodyWriter extends VertxAsyncFileMessageBodyWriter
+        implements ServerMessageBodyWriter<AsyncFile> {
+
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, ResteasyReactiveResourceInfo target, MediaType mediaType) {
+        // allow for subtypes, such as AsyncFileImpl
+        return AsyncFile.class.isAssignableFrom(type);
+    }
+
+    @Override
+    public void writeResponse(AsyncFile file, Type genericType, ServerRequestContext context) throws WebApplicationException {
+        ResteasyReactiveRequestContext ctx = ((ResteasyReactiveRequestContext) context);
+        ctx.suspend();
+        ServerHttpResponse response = context.serverResponse();
+        response.setChunked(true);
+        file.handler(buffer -> {
+            try {
+                response.write(buffer.getBytes());
+            } catch (Exception x) {
+                // believe it or not, this throws
+                ctx.resume(x);
+                return;
+            }
+            if (response.isWriteQueueFull()) {
+                file.pause();
+                response.addDrainHandler(new Runnable() {
+                    @Override
+                    public void run() {
+                        file.resume();
+                    }
+                });
+            }
+        });
+
+        file.endHandler(new Handler<Void>() {
+            @Override
+            public void handle(Void event) {
+                response.end();
+                // Not sure if I need to resume, actually
+                ctx.resume();
```

Comment:
Likely not, since there should be no more handlers in the chain, right?

Reply:
Well, there's always cleanup, no?

Reply:
What cleanup? I mean if this is the last Handler in the chain, resuming should do nothing but push the context into the event loop and just return when the event actually gets processed

Reply:
In the context.close we do things like call async listeners and tear down the CDI scope and stuff, no?

Reply:
Ah, good point